cds/modules/search/access_control.py

```diff
@@ -0,0 +1,92 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2016 CERN.
+#
+# Invenio is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Invenio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the Free Software Foundation, Inc.,
+# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+""" Search access control. """
+
+from elasticsearch_dsl.query import Q
+from flask import g
+from invenio_records.api import Record
+from invenio_search import RecordsSearch
+from invenio_search.api import DefaultFilter
+
+
+def cern_read_factory(record, *args, **kwargs):
+    """Restrict search results based on CERN groups and user e-mail.
+
+     Records should have an '_access' field containing three nested fields,
+     namely 'read', 'write' and 'admin', each having zero or more groups/emails
+     that are given the enclosing access right.
+
+     JSON example:
+
+      "_access": [
+         {
+          "read": ["it-dep-cda", "it-dep", "reader@cern.ch"],
+          "write": [], # ANYONE can write!!
+          "admin": ["orestis.melkon@cern.ch"]
+         }
+      ]
+    """
+
+    def can(self):
+        """Cross-check user's CERN groups with the record's '_access' field."""
+
+        user_groups = [str(need.value) for need in sorted(g.identity.provides)]
+
+        rec = Record.get_record(record.id)
+
+        # Records with no `_access` field are public
+        if '_access' not in rec:
+            return True
+
+        rec_groups = rec['_access']['read']
```

Comment:
I don't know exactly what you mean by `read = read + write + admin`.
Probably we will drop `write` and `admin` keys from there as they probably live in other place, i.e. deposit
Also we might need to split the read one into two, one for the _system_ rights and the other for the _user_ rights to avoid problems when doing updates